services/web/server/src/simcore_service_webserver/projects/_folders_api.py

```diff
@@ -1,69 +1,87 @@
-            connection,
-            project_uuid=project_id,
-        )
-        if _source_folder_id is None:
-            assert folder_id is not None  # nosec
-            # NOTE: folder permissions are checked inside the function
-            await folders_db.folder_add_project(
-                connection,
-                product_name=product_name,
-                folder_id=folder_id,
-                gids={user["primary_gid"]},
+    # In personal workspace user can move as he wish, but in the
+    # shared workspace user needs to have write permission
+    _personal_workspace_user_id_or_none: UserID | None = user_id
+    if project_db.workspace_id is not None:  # shared workspace
+        if project_access_rights.write is False:
+            raise ProjectInvalidRightsError(
+                user_id=user_id,
                 project_uuid=project_id,
+                reason=f"User does not have write access to project {project_id}",
             )
-            return
+        # Setup _personal_workspace_user_id_or_none to None, as this is not a personal workspace
+        _personal_workspace_user_id_or_none = None
 
-        # NOTE: folder permissions are checked inside the function
-        await folders_db.folder_move_project(
-            connection,
+    if folder_id:
+        # Check user has access to folder
+        await folders_db.get_for_user_or_workspace(
+            app,
+            folder_id=folder_id,
             product_name=product_name,
-            source_folder_id=_source_folder_id,
-            gids={user["primary_gid"]},
-            project_uuid=project_id,
-            destination_folder_id=folder_id,
+            user_id=_personal_workspace_user_id_or_none,
+            workspace_id=project_db.workspace_id,
         )
-        return
+
+    # Move project to folder
+    prj_to_folder_db = await project_to_folders_db.get_project_to_folder(
+        app,
+        project_id=project_id,
+        personal_workspace_user_id_or_none=_personal_workspace_user_id_or_none,
+    )
+    if prj_to_folder_db is None:
+        if folder_id is None:
+            return
+        await project_to_folders_db.insert_project_to_folder(
+            app,
+            project_id=project_id,
+            folder_id=folder_id,
+            personal_workspace_user_id_or_none=_personal_workspace_user_id_or_none,
+        )
+    else:
+        # Delete old
+        await project_to_folders_db.delete_project_to_folder(
+            app,
+            project_id=project_id,
+            folder_id=prj_to_folder_db.folder_id,
+            personal_workspace_user_id_or_none=_personal_workspace_user_id_or_none,
+        )
+        # Create new
+        if folder_id is not None:
+            await project_to_folders_db.insert_project_to_folder(
+                app,
+                project_id=project_id,
+                folder_id=folder_id,
+                personal_workspace_user_id_or_none=_personal_workspace_user_id_or_none,
+            )
```

Comment:
this is not atomic. It will create an issue at some point!
For example the DB goes down on the second operation. Instead of moving you have deleted a folder.

Reply:
Yes I know, as always I do not want to overoptimize, for now its doing its job.